Front end rework

benmcmahon100 opened the pull request:
Passes the gulp test-challenges
Closes #3
Closes #4 
Closes #18 
Closes #19 
Closes #20 
Closes #22 
Closes #31 
Closes #33 
Closes #34 
Closes #35

benmcmahon100 committed:
New challenge creation fixed

benmcmahon100 committed:
Challenge Duping added and missing fields are added to challenges upon loading. Click box for opening challenges has been slightly reduced

benmcmahon100 committed:
Tabbed Editor and Custom Code Mirror sizes implemented. Misc tab has the odd behaviour of having all the un-rendered code mirrors instead of the ones that aren't defined in the layout. These should be avoided as they won't update the seed data on a save. Also added a nice font to the other elements

benmcmahon100 committed:
Layout and challenge skeleton now loaded via configs, new challenge appears under new challenge button and duped challenges appear under their parent

benmcmahon100 committed:
z-index fix

benmcmahon100 committed:
Code highlighting should be fixed.
Home and End key bindings are now implemented.

benmcmahon100 committed:
Hide broken misc tab

benmcmahon100 committed:
Order of challenge Properties is now standardised

benmcmahon100 committed:
Merge remote-tracking branch 'remotes/origin/master' into FrontEndRework

SaintPeter commented on app.js:
```diff
@@ -27,6 +27,46 @@ app.post('/export', function(req, res, next) {
+      keys.forEach(function(key){
+        newData[key] = challenge[key];
+      });
+      console.log(keys);
```
Hello Mr. Console.log!

SaintPeter commented on public/editorLayout.json:
```diff
@@ -0,0 +1,23 @@
+  "descriptionEs": "",
+  "namePt": "",
+  "descriptionPt": ""
+}
```
Newlines?  We don't need no steenking Newlines!

SaintPeter commented on src/Components/Editor.js:
```diff
@@ -75,18 +88,51 @@ class Editor extends Component {
       );
     });
 
+    for(var i = 0; i < unrenderedCodeMirrors.length; i++){
```
Why not forEach?

SaintPeter commented on src/Components/GrandCentralStation.js:
```diff
@@ -230,6 +226,45 @@ class GrandCentralStation extends Component {
+            }
+          }
+
+          console.log(index);
```
Hello Mr. Console.log :wink:

SaintPeter commented:
The Prev and Next buttons break the app.

SaintPeter commented:
Also, some numeric fields are converted to strings:
challengeType - in the challenge
order - at the top level of the file

benmcmahon100 committed:
Fixed the prev & next buttons.
Fixed config names

benmcmahon100 committed:
Visual tweaks
Tracked down codemirrors' value of null error
small misc bug fixes
added mini-docs
Fixed ints being parsed as strings

SaintPeter committed:
Fix order being zeroed

SaintPeter commented:
:+1:
